src/service/account.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * # Account Service
+ *
+ * For account sign in & sign up.
+ *
+ * ## Injected Dependency
+ *
+ * - [UserModel](../classes/model_user.UserModel.html)
+ *
+ * @module
+ */
+import assert from 'assert';
+import { UserModel } from '../model/user';
+import { User } from '../entity/user';
+import { ERROR } from '../config/constant';
+import { encodeWithSalt } from '../util/crypto';
+import { Inject, Provide } from '../util/container';
+
+@Provide()
+export class AccountService {
+  @Inject()
+  private userModel: UserModel;
+
+  /**
+   * Sign in with email & password
+   * @param email User email
+   * @param password User password
+   * @returns entire user entity
+   */
+  async signIn(email: string, password: string) {
+    // 1. Find user
+    const user = await this.userModel.getOneByEmail(email);
+    assert(!!user, ERROR.SERVICE_ACCOUNT_SIGNIN_NOTFOUND_EMAIL);
+
+    // 2. Verify password
+    if (user.password !== this.encodePassword(user, password)) {
+      // Could count+1 for count > 2 block the sign in
+      return Promise.reject(ERROR.SERVICE_ACCOUNT_SIGNIN_PASSWORD);
```

Comment:
why not `throw ERROR.SERVICE_ACCOUNT_SIGNIN_PASSWORD` here, is it a style preference?

Reply:
Um, when I first wrote about that, it looked like:

```ts
assert(
  user.password === this.encodePassword(user, password),
  ERROR.SERVICE_ACCOUNT_SIGNIN_PASSWORD,
)
```

What came to me later, is that I tought there might be more requirements like  frequent failed events blocking, thus why I turned it into if-else block, and in my option, inside this if-code-block should be a Promise (might be return from another function), so I decided to use a `Promise.rejection` as a Promise placeholder.